File: java/138.copy-list-with-random-pointer.java
/*
 * @lc app=leetcode id=138 lang=java
 *
 * [138] Copy List with Random Pointer
 */

// @lc code=start
/*
// Definition for a Node.
class Node {
    int val;
    Node next;
    Node random;

    public Node(int val) {
        this.val = val;
        this.next = null;
        this.random = null;
    }
}
*/

class Solution {
    public Node copyRandomList(Node head) {
        Node dummyHead = new Node(0);
        Node curr, next;
        curr = dummyHead;

        HashMap<Node, Node> map = new HashMap();

        while(head != null) {
            next = new Node(head.val);
            curr.next = next;
            next.random = head.random;

            map.put(head, next);

            head = head.next;
            curr = next;
        }

        curr = dummyHead.next;
        while(curr != null) {
            curr.random = map.getOrDefault(curr.random, null);

            curr = curr.next;
        }

        return dummyHead.next;
    }
}
// @lc code=end

File: cpp/138.copy-list-with-random-pointer.cpp
/*
 * @lc app=leetcode id=138 lang=cpp
 *
 * [138] Copy List with Random Pointer
 */

// @lc code=start
/*
// Definition for a Node.
class Node {
public:
    int val;
    Node* next;
    Node* random;
    
    Node(int _val) {
        val = _val;
        next = NULL;
        random = NULL;
    }
};
*/

class Solution {
public:
    Node* copyRandomList(Node* head) {
        std::unordered_map<Node*, Node*> map;
        
        Node dummy(0);

        Node * curr, * prev;

        prev = &dummy;

        while(head) {
            curr = new Node(head->val);

            map[head] = curr;
            curr->random = head->random;

            prev->next = curr;
            prev = curr;

            head = head->next;
        }

        curr = dummy.next;

        while(curr) {
            curr->random = map[curr->random];

            curr = curr->next;
        }

        return dummy.next;
    }
};
// @lc code=end

